haystack/pipelines/config.py

```diff
@@ -8,23 +8,24 @@
 from pathlib import Path
 
 import yaml
-from networkx import DiGraph
+import networkx as nx
 from jsonschema.validators import Draft7Validator
 from jsonschema.exceptions import ValidationError
 
 from haystack import __version__
-from haystack.nodes.base import BaseComponent
+from haystack.nodes.base import BaseComponent, RootNode
 from haystack.nodes._json_schema import inject_definition_in_schema, JSON_SCHEMAS_PATH
-from haystack.errors import PipelineConfigError, PipelineSchemaError
+from haystack.errors import PipelineError, PipelineConfigError, PipelineSchemaError
 
 
 logger = logging.getLogger(__name__)
 
 
-VALID_INPUT_REGEX = re.compile(r"^[-a-zA-Z0-9_/.:]+$")
+VALID_INPUT_REGEX = re.compile(r"^[-a-zA-Z0-9_/\\.:]+$")
+VALID_ROOT_NODES = ["Query", "File"]
 
 
-def get_pipeline_definition(pipeline_config: Dict[str, Any], pipeline_name: Optional[str] = None) -> Dict[str, Any]:
+def get_pipeline_definition(config: Dict[str, Any], pipeline_name: Optional[str] = None) -> Dict[str, Any]:
     """
     Get the definition of Pipeline from a given pipeline config. If the config contains more than one Pipeline,
```

Comment:
Let's briefly talk about the advantages and disadvantages of renaming pipeline_config to config in the code. For example, I see that the doc string mentions here "pipeline config", which seems better to me as it explicitly mentions that the config is about a pipeline (or in fact multiple pipelines).

Reply:
I renamed this parameter out of frustration at the n-th time that I got confused between `pipeline_definition` and `pipeline_config`, so first of all sorry for the sudden change :smile: 

On the other hand I believe this concept needs renaming. `pipeline_config` is simply misleading when we have `component_configuration` and `pipeline_definition` in the codebase, and it's not very descriptive of what it is, At _least_ is should be plural, so `pipelineS_config`, although that won't help much when coding. So for now I went for `config` (to match the `get_config()` method) to see how it would look in the code.

I agree `config` is not nice either, but let's take the chance of these breaking changes and rename this parameter. I believe in DC they're having the same conceptual issue, so a decision on our end might benefit them too.

Other alternatives I came up with:

- `system_config`: too generic?
- `yaml_config`: too specific? (although I see little use of the config outside of loading from YAML, but I might be wrong)
- `pipelines_and_nodes_config`: too long?
- `complete_config`: vague?
- `complete_pipeline_config`: still vague _and_ long
- `pipeline_group_config`, `pipeline_group_definition`, `pipeline_group_description`: a bit long too?
- Other ideas?

We could also agree on a name for a group of pipelines and stick with it (like `system`, `stack`, `bundle`, `plumbing` or whatever collective group for pipes exists in English :smile: ). Indeed `stack_config` (or `bundle_config`) doesn't sound too bad if we clarify in the code that `stack` (or `bundle`) is a group of Pipelines...